proposals/0014-wave-size-range.md

```diff
@@ -0,0 +1,320 @@
+ `maxWaveSize` is the maximum wave size supported by the shader
+ representing the end of the allowed range,
+ and `prefWaveSize` is the optional preferred wave size representing the size
+ expected to be the most optimal for this shader.
+
+### DXIL Additions
+
+The existing `WaveSize` value is stored as a metadata 32-bit integer constant.
+To store the additional values, an additional metadata tag indicating a tuple of
+ three 32-bit integers representing the given parameters to `WaveSize`.
+In the case where no preferred wave size is specified, the value zero will
+ indicate that nothing was specified.
+
+|         Tag             | Constant |         Value           |
+|-------------------------|----------|-------------------------|
+|kDxilRangedWaveSizeTag   |    23    |MD list: (i32, i32, i32) |
+
+### SPIR-V Additions
+
+To represent the wave size range and preferred value in SPIR-V,
+ a new `ExecutionMode` value would need to be defined for the `OpExecutionMode`
+ instruction to allow specifying `SubgroupSize` with three operands instead of
+ one.
+Like the other formats, the location of the operands would indicate
+ what each value represents, the first two being the min and max wave sizes and
+ the third being the preferred wave size.
+
+### Diagnostic Changes
+
+#### New Errors
+
+These are where new or slightly altered errors are produced:
+
+* If any of the parameters of `WaveSize` are not literal power of two integers
+  between 4 and 128 inclusive.
+* If the minimum wave size value is greater than or equal to the max wave size
+ as respectively specified by the first and second parameters in the `WaveSize`
+ attribute.
+* If the preferred wave size value indicated by the third parameter of
+ `WaveSize` is not in the range specified by the first two parameters.
+* If multiple `WaveSize` attributes are applied to the same entry point
+ with different numbers or values of parameters,
+ the existing error indicating an attribute conflict is produced.
+* If more than three or fewer than one parameter is applied to `WaveSize`.
+* If negative values are provided for any of the `WaveSize` parameters.
+* If float values are provided for any of the `WaveSize` parameters.
+* If non-numerical values are provided for any of the `WaveSize` parameters.
+* If non-literal variables are provided for any of the `WaveSize` parameters.
+* If any of the parameters to `WaveSize` are less than 4 or greater than 128.
+
+#### Validation Changes
+
+Validation should confirm:
+
+* Each element in that is a power of two integer between 4 and 128 inclusive.
+  The third parameter may also be zero.
+* The minimum wave size is less than the maximum wave size.
+* The preferred wave size is zero or lies between the minimum and maximum.
+* Where validator version is less than 1.6, fail on shaders that use `WaveSize`.
+* Where validator version is less than 1.8, fail when the `WaveSize` metadata
+ tag has more than one parameter.
```

Comment:
```suggestion
* Where validator version is less than 1.7, fail on shaders that use `RangedWaveSize`.
* Validation should fail if an entry has both `RangedWaveSize` and `WaveSize` metadata.
```
Since we are using different tags for ranged and the old wave size, the validation rules should be simpler.

Reply:
I think this should be where the validator is less than 1.8.

I included the mention of `WaveSize` because even though it's a different spec, it seems we might, with minimal effort correct some testing holes in the process. Perhaps that would be better done with individual bugs though.

Reply:
This still needs correction.